bionic/code_references.py

```diff
@@ -0,0 +1,191 @@
+
+    Referenced objects can be anything from a class, a function, a module or any
+    variables. The references can also be any scope, like global, local, cell,
+    free, or can even be in another object referenced using a full qualified
+    name.
+
+    Note that this function does not perform any runtime logic. This includes
+    calling any functions used by the input code because doing so can be
+    expensive and have unintended consequences. Due to this, any references with
+    a fully qualified name that uses result of such a call won't be detected.
+    This means that if an inner function returns a module or a class, any
+    variables of the module or the class won't be detected. In this case, we will
+    return the name of the reference as a proxy. So for a function like below,
+    the returned reference would be ``"call"`` instead of ``MyClass.call``.
+
+    .. code-block:: python
+
+        def x():
+            my_cls = get_my_class("MyClass") # Returns class MyClass
+            my_cls.call()
+
+
+    On a higher level, for Smart Caching, this leaves some changes that Bionic
+    won't be able to detect. Bionic should be able to detect any changes except
+    when a function returns a class or a module. In those case, Bionic cannot
+    detect what is returned. But it should be able to detect the class / module
+    returned when hashing the function which returns the said result. Bionic will
+    still hash those classes / modules, and detect as many changes as it can in
+    them.
+    """
+    # Top of the stack.
+    tos = None
+    lineno = None
+    refs = []
+
+    def set_tos(t):
+        nonlocal tos
+        if tos is not None:
+            # If the top of stack item already exists, that means we
+            # have gone through all the instructions that use the item,
+            # and it is a reference object.
+            refs.append(tos)
+        tos = t
+
+    # Our goal is to find referenced objects. The problem is that co_names
+    # does not have full qualified names in it. So if you access `foo.bar`,
+    # co_names has `foo` and `bar` in it but it doesn't tell us that the
+    # code reads `bar` of `foo`. We are going over the bytecode to resolve
+    # from which object an attribute is requested.
+    # Read more about bytecode at https://docs.python.org/3/library/dis.html
+
+    for op in dis.get_instructions(code):
+        try:
+            # Sometimes starts_line is None, in which case let's just remember the
+            # previous start_line (if any). This way when there's an exception we at
+            # least can point users somewhat near the line where the error stems from.
+            if op.starts_line is not None:
+                lineno = op.starts_line
+
+            if op.opname in ["LOAD_GLOBAL", "LOAD_NAME"]:
+                if op.argval in context.globals:
+                    set_tos(context.globals[op.argval])
+                else:
+                    set_tos(op.argval)
```

Comment:
Thinking about this more: here `op.argval` is just a string, right? So if later we have a `LOAD_ATTR` instruction, we will call `getattr` on this string? Is that what we want?

Reply:
Also, it might be helpful to have a comment explaining how this case arises. If we see a `LOAD_NAME` and it's not in the global names, what other places could it be in?

Reply:
If we get `LOAD_ATTR` when `TOS` is string, we concatenate the TOS with the attribute name.

From my understanding of Python and bytecode, I think this can only happen if the variable does not exist. If you look at the test `test_function_references`, you'll see that the last example in that test return a `"func_does_not_exist"` string because the function does not exist and cannot be found.